New structure for predictors

marcinpaluch1994 opened the pull request:
Flexible specification of predictors to be used:

# predictors_specifications may be a names from the list of predictors in config_predictors.yml (assuming specification make sense)
# It can be a names of networks or GPs e.g. 'GRU-6IN-32H1-32H2-5OUT-0' to be used for test
# For multiple network they must be in the same folder specified in predictors:neural_default:path_to_model
# One can also specify full path to each model separately as model_name,
# in this case the path from predictors:neural_default:path_to_model get overwritten
# Where multiple ODE models are available, one can specify the name after colon e.g. ["ODE_TF:simple", "ODE:st"]
# In this case name from predictor_config is getting overwritten

The PredictorWrapper API is fully compatible with old and new ideas:
        self.predictor = PredictorWrapper()
        self.predictor.configure(batch_size=self.num_rollouts, horizon=self.cem_samples, predictor_specification=predictor_specification)

Currently self.predictor = PredictorWrapper() is within controllers, but it can be pushed outside and predictor passed as an argument to a controller. In the last case also predictor.update can be pushed to the outside of the controller
self.predictor.configure takes predictor_specification as an optional argument. If provided the choice of predictor is coming from controller. If None/ not used the default predictor as specified in predictors config 
The config is loaded by PredictorWrapper, but it can also without additional changes be passed to the PredictorWrapper.__init__()

Knowen Issues:
Many controllers in Cartpole GUI are not working, but I don’t think it is because of predictor changes. They mostly complain about missing parameters from config I was not removing.


TODO:
It would be high time to time the predictors and see if they have still the same speed as half a year ago.

Last comment:
Could the cost and also the  controllers have (almost) exactly same structure? With config_costs and config_controllers in Control_Toolkit, cost being set either in config or in controller, controller being set either in its config or in Planner?

marcinpaluch1994 committed:
PredictorWrapper and respective changes in Brunton Tests. Brunton Tests works, other things not yet

marcinpaluch1994 committed:
updating predictors

marcinpaluch1994 committed:
Updating GUI

marcinpaluch1994 committed:
Yet new structure for predictors

marcinpaluch1994 committed:
New structure of predictors now works for Brunton Plots and for (many of) ODE_TF based controllers.

frehe commented on SI_Toolkit_ASF/config_predictors.yml:
```diff
@@ -0,0 +1,40 @@
+predictor_name_main: 'I_love_control'
```
:)

frehe commented on SI_Toolkit_ASF/config_predictors.yml:
```diff
@@ -0,0 +1,40 @@
+
+  # ADD YOUR PREDICTORS BELOW
+
+  I_love_control:
+    predictor_type: 'neural'  # Possible options are: 'neural', 'GP', 'ODE, 'ODE_TF'
+    model_name: 'GRU-6IN-32H1-32H2-5OUT-0'
+    path_to_model: './SI_Toolkit_ASF/Experiments/Pretrained-RNN-1/Models/'
+    dt: # Matters only for differential network
+    update_before_predicting: True
+
+  I_love_control_too:
+    predictor_type: 'ODE'
+    model_name:
+    dt: 0.02
+    intermediate_steps: 10
```
I absolutely love this structure

frehe commented on SI_Toolkit_ASF/config_predictors.yml:
```diff
@@ -0,0 +1,40 @@
+    dt: 0.02
+    intermediate_steps: 5
+
+  # ADD YOUR PREDICTORS BELOW
+
+  I_love_control:
+    predictor_type: 'neural'  # Possible options are: 'neural', 'GP', 'ODE, 'ODE_TF'
+    model_name: 'GRU-6IN-32H1-32H2-5OUT-0'
+    path_to_model: './SI_Toolkit_ASF/Experiments/Pretrained-RNN-1/Models/'
+    dt: # Matters only for differential network
+    update_before_predicting: True
+
+  I_love_control_too:
+    predictor_type: 'ODE'
+    model_name:
+    dt: 0.02
+    intermediate_steps: 10
```
Not seeing any info on how to use predictor_specification here. In config.yml it's sort of clear

marcinpaluch1994 replied:
Added at the top. Is it what you wanted to see?

frehe replied:
Yes, got it now

frehe commented on config.yml:
```diff
@@ -30,7 +26,7 @@ controller:
     dt: 0.02                              # sec
     mpc_horizon: 35                       # steps
-    CEM_NET_NAME: GRU-3IN-32H1-32H2-2OUT-0
+    predictor_specification: "ODE_TF"    # Can be "ODE", "ODE_TF", network/GP name (possibly with path) e.g. 'GRU-6IN-32H1-32H2-5OUT-0' or a name of a custom predictor. For more info see config_predictors in SI_Toolkit_ASF
```
How would an example spec for the RNN predictor with custom network look?

marcinpaluch1994 replied:
Not quite getting a question... You either specify the network name if it is "our" custom type. Then you can neural_default from config_predictors with model_name (and possibly path, if you give here a full path) overwritten. Or you need to reference here a completely different predictor by name (and it must be listed in config_predictors) and PredictorWrapper must take care on how to load such a predictor.

frehe commented on config.yml:
```diff
@@ -272,5 +252,5 @@ cartpole:
 
 
-data_generator:
+data_generator: # FIXME: Warum seed for data_generator here?
   seed: 1  # If null random seed based on datetime is used
```
This can be removed. Then, only need to edit the `run_data_generator_legacy.py` and load the seed from `config_data_gen.yml`. (in case we want to retain compatibility)

marcinpaluch1994 replied:
Implemented! I am not testing run_data_generator_legacy.py though. It should be fine now, otherwise it is of little relevance I guess... Normal data generator tested & working.

marcinpaluch1994 committed:
Removing data_generator:seed from general config.yml

marcinpaluch1994 committed:
Changing predictor_name_main to predictor_name_default and adding explanation

marcinpaluch1994 committed:
adding explanations

frehe commented on others/Diego_semester_project/run_data_generator_legacy.py:
```diff
@@ -18,10 +18,11 @@
 
 def run_data_generator(run_for_ML_Pipeline=False, record_path=None):
-    seed = config_CartPole["data_generator"]["seed"]
+    seed = config_data_gen["data_generator"]["seed"]
```
```suggestion
    seed = config_data_gen["seed"]
```

frehe replied:
```suggestion
    seed = config_data_gen["seed"]
```

marcinpaluch1994 committed:
Moving the computation_library.py to SI_Toolkit. SI_Toolkit now independent of Contorl Toolkit. Might me messing up Control Gym and (less likely) f1t

marcinpaluch1994 committed:
Updating reference to subrepos

marcinpaluch1994 committed:
Bug in run_data_generator_legacy.py corrected

frehe commented:
Could we merge my part into this feature branch, then merge into master?

marcinpaluch1994 commented:
Yes! Please!

frehe committed:
Update refs to submodules

frehe committed:
Update refs

frehe committed:
Update submodule

frehe committed:
Restructuring Work (#40)

* Update reference to control toolkit

* - Gym Env Wrapper: Plot target position if available

* Update dist_adam_resamp2 name to rpgd

* Update controller name to RPGD

* Changes related to upgrade to gym>=0.26

* - Controllers take predictor instead of true env

* - Use correct variable to reference render mode

* - Adopt new predictor structure in CartPole class
- Update next step predictors
- Ensure the training -> predictor_autoregressive_tf pipeline works

* Update refs

* Some updated type hints

* Export state variables in ASF

* Make GUI work

* - Move cost functions to ASF

* Update refs

* Adopt the new conventions in the ASF controllers

* - Move controller configuration to its own config in Control_Toolkit_ASF
- Remove target attributes from CartPole class (only in cost function now)
- Move controller/cost combination into a parameter-free planner you can instantiate anywhere
- Add an own cost function config
- Remove predictor specification from the ASF controllers that don't need a SI_Toolkit predictor
- cost function has a computation library now
- move cost functions to Control_Toolkit_ASF

* minimal config change

* Remove step_fun references, moved computation lib to SI_Toolkit

* Update refs to submodules

* - Display selectable combination of controller and optimizer in GUI
- Instantiate and configure the controller from CartPole.set_controller
- Separate controller __init__ (same for all controllers) and configuration (happens after creation)
- "mppi" controller is now "mppi_cartpole"
- When logging enabled, optimizer classes send logs to controller (simplified observer pattern)
- Cost function now takes a controller reference, and a computation library to init
- Controllers have explicit _computation_library (set as class variable, private, read-only)
-

* Fix tests

* Update controller_secloc.py

This should not be a task of controller to simulate effect of noise. I remove it.

* - Make controller MPC TF into general MPC controller
- Controllers now take number of states/actions and action limits instead of gym spaces

* Remove no longer required imports from ASF controllers

* Move base cost function to control toolkit

* Use cost function wrapper

* Move ASF controller descriptions to ASF files

* Inherit get_trajectory_cost from parent

* Updated ref

* - Remove cost function from cartpole config
- Update ref
- Add predictor specification to controller config

Co-authored-by: Frederik Heetmeyer <heetmeyf@ethz.ch>
Co-authored-by: Marcin Paluch <62556049+marcinpaluch1994@users.noreply.github.com>

marcinpaluch1994 committed:
same way of using model_name and path to model in predictor_autoregressive_GP.py as in predictor_autoregressive_tf.py

marcinpaluch1994 committed:
Adding double quotes to type[...] to make it compatible with Python 3.8

frehe committed:
Update Control Toolkit reference

marcinpaluch1994 committed:
Broken! Reworking loading of the config controllers.

frehe committed:
Make controller configuration an instance attribute
Remove yaml load of controller config file from all controllers (except 2 ASF controllers where the parameters are global in module scope)

frehe committed:
Update from toolkits

frehe committed:
Update from toolkits

frehe committed:
Update Toolkit

marcinpaluch1994 committed:
Added support for RNNs in predictor_autoregressive_tf.py. From now on predictor_autoregressive_tf.py is is fully functional for both TF and Pytorch and can be renamed.

marcinpaluch1994 committed:
Changing default intermediate_steps for ODE_TF for predictions to coincide perfectly with cartpole simulator.

marcinpaluch1994 committed:
More information added to config.

marcinpaluch1994 committed:
Renaming predictor_autoregressive_tf.py to predictor_autoregressive_neural.py!

marcinpaluch1994 committed:
Big cleaning in GPs!
Tested: Training, Predictor and Timing

marcinpaluch1994 committed:
Merge remote-tracking branch 'origin/master' into Marcin_restructurisation
